File: week3/joehyun000/BOJ_1914.cpp
#include <iostream>
#include <cmath>
using namespace std;

void hanoi(int n, int start, int end, int mid) {
    if (n == 1) {
        printf("%d %d\n", start, end);
        return;
    }
    hanoi(n - 1, start, mid, end);
    printf("%d %d\n", start, end);
    hanoi(n - 1, mid, end, start);
}

int main() {
    int N;
    cin >> N;

    cout << (1LL << (long long)N) - 1 << endl; 

    if (N <= 20) {
        hanoi(N, 1, 3, 2, (1LL << N) - 1);
    } else {
        hanoi(N, 1, 3, 2, 20);
    }

    return 0;
}

File: week3/Jhwan02/BOJ1914.py
def hanoi(n,s,e):
    if n == 1:
        print(s,e)
        return

    hanoi(n-1,s,6-s-e)
    print(s,e)
    hanoi(n-1,6-s-e,e)



n = int(input())

print(2**n-1)
if n<=20:
    hanoi(n,1,3)